### DISCUSSION TITLE
Can you throw/return errors inside a DataFetchingEnvironment

### DISCUSSION THREAD

**Discussion by ms185730** [graphql-java/graphql-java#2443] [General]
Hello, 

I have a dataFetcher that looks like this: 

```
public DataFetcher<Group> getGroup() {
        return dataFetchingEnvironment -> {

            final String groupId = dataFetchingEnvironment.getArgument(Constants.ARG_ID);

            final var group = groupRepository.findById(groupId, org);

            if (group.isEmpty()) {
                throw new ResourceExceptionBuilder.buildDoesNotExistException(new GroupIdData(groupId));
            }

            return group.get();
        };
    }
```

If there is no group with the provided groupId, I'd like to throw a ResourceDoesNotExist exception, but I'm having trouble returning or throwing the error. Is it possible to do so in a dataFetcher like this?

**Comment by bbakerman** [comment]
> but I'm having trouble returning or throwing the error. Is it possible to do so in a dataFetcher like this?

Its totally possible.  It's not recommended but its possible.  When you say you are having trouble do it,  can you please expand that more.

GraphQL uses an error shape to indicate problems with fields fetchers.  If you throw and exception our default `graphql.execution.SimpleDataFetcherExceptionHandler` will turn it into a graphql error for you.

There is also `graphql.execution.DataFetcherResult` as a mechanism to allow both data and errors (or either) to be returned

**Reply by ms185730** [reply]
If I have
`throw new Exception();`
I get a compiler error: `java: unreported exception java.lang.Exception; must be caught or declared to be thrown`
And when I put `throws Exception` in the method header, I still get the same compiler message.
I can't return the exception because of the type mismatch. Am I only allowed to throw GraphQL errors/exceptions?

**Reply by bbakerman** [reply]
This is a basic a Java question not so much a graphql-java one

https://www.baeldung.com/java-checked-unchecked-exception

```
DataFetcher df = env -> {
   if (env.getArgument("x").equals("y") {
      throw new RuntimeException("The will work - its an un checked exception")
}
```
